pkg/controller/gcp/database/instance.go

```diff
@@ -0,0 +1,236 @@
+		return requeueNever, handleNotFound(err)
+	}
+
+	// create local operations to handle Kubernetes (local) types operations
+	lops := r.factory.makeLocalOperations(i, r.Client)
+
+	// create managed operations to handle CloudSQL (managed) instance operations
+	mops, err := r.factory.makeManagedOperations(ctx, i, lops)
+	if err != nil {
+		return requeueNow, lops.updateReconcileStatus(ctx, err)
+	}
+
+	// create syncdeleter to handle top level reconciliation operations
+	sd := r.factory.makeSyncDeleter(mops)
+
+	if meta.WasDeleted(i) {
+		return sd.delete(ctx)
+	}
+
+	return sd.sync(ctx)
+}
+
+type factory interface {
+	makeLocalOperations(*v1alpha1.CloudsqlInstance, client.Client) localOperations
+	makeManagedOperations(context.Context, *v1alpha1.CloudsqlInstance, localOperations) (managedOperations, error)
+	makeSyncDeleter(managedOperations) syncdeleter
+}
+
+type operationsFactory struct {
+	client.Client
+}
+
+var _ factory = &operationsFactory{}
+
+func (f *operationsFactory) makeLocalOperations(inst *v1alpha1.CloudsqlInstance, kube client.Client) localOperations {
+	return newLocalHandler(inst, kube)
+}
+
+func (f *operationsFactory) makeManagedOperations(ctx context.Context, inst *v1alpha1.CloudsqlInstance, ops localOperations) (managedOperations, error) {
+	p := &gcpv1alpha1.Provider{}
+	n := meta.NamespacedNameOf(inst.GetProviderReference())
+	if err := f.Get(ctx, n, p); err != nil {
+		return nil, err
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.GetNamespace(), Name: p.Spec.Secret.Name}
+	if err := f.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider's secret %s", n)
+	}
+
+	creds, err := google.CredentialsFromJSON(context.Background(), s.Data[p.Spec.Secret.Key], cloudsql.DefaultScope)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot retrieve creds from json")
+	}
+
+	return newManagedHandler(ctx, inst, ops, creds)
+}
+
+func (f *operationsFactory) makeSyncDeleter(ops managedOperations) syncdeleter {
+	return &instanceSyncDeleter{
+		managedOperations: ops,
+		createupdater:     &instanceCreateUpdater{managedOperations: ops},
+	}
+}
+
+type syncdeleter interface {
+	sync(context.Context) (reconcile.Result, error)
+	delete(context.Context) (reconcile.Result, error)
+}
+
+type instanceSyncDeleter struct {
+	managedOperations
+	createupdater
+}
+
+func (sd *instanceSyncDeleter) delete(ctx context.Context) (reconcile.Result, error) {
+	if sd.isReclaimDelete() {
+		if err := handleNotFound(sd.deleteInstance(ctx)); err != nil {
+			return requeueNow, sd.updateReconcileStatus(ctx, err)
+		}
+	}
+	return requeueNow, sd.removeFinalizer(ctx)
+}
+
+// sync - synchronizes the state of the cloudsql instance instance with the
+// state of the obj object
+func (sd *instanceSyncDeleter) sync(ctx context.Context) (reconcile.Result, error) {
+	inst, err := sd.getInstance(ctx)
+	if err != nil && !googleapi.IsErrorNotFound(err) {
```

Comment:
```suggestion
	if resource.Ignore(googleapi.IsErrorNotFound, err) != nil {
```

Reply:
That's fine. TBH: I think I prefer the original.